recipes/cpr/all/conanfile.py

```diff
@@ -34,38 +55,64 @@ def config_options(self):
     def configure(self):
         if self.options.shared:
             del self.options.fPIC
+        if not self._supports_openssl:
+            del self.options.with_openssl
+        if not self._supports_winssl:
+            del self.options.with_winssl
+
+        if self._supports_openssl and tools.is_apple_os(self.settings.os):
+            self.options.with_openssl = False # Default libcurl in CCI is `with_ssl="darwin"` which is unclear if cpr supports this
+            
+        # Make sure libcurl uses the same SSL implementation
+        if self.options.get_safe("with_openssl", False):
+            # self.options["libcurl"].with_openssl = True # deprecated in https://github.com/conan-io/conan-center-index/pull/2880
+            self.options["libcurl"].with_ssl = "openssl"
+        if self.options.get_safe("with_winssl", False):
+            # self.options["libcurl"].with_winssl = True # deprecated in https://github.com/conan-io/conan-center-index/pull/2880
+            self.options["libcurl"].with_ssl = "schannel"
+
+        if self.options.get_safe("with_winssl", False) and self.settings.os != "Windows":
+            raise ConanInvalidConfiguration("cpr only supports winssl on Windows")
+
+        if self.options.get_safe("with_openssl", False) and self.options.get_safe("with_winssl", False):
+            raise ConanInvalidConfiguration("cpr can not be built with both openssl and winssl")
+            
+        if self.settings.compiler == "Visual Studio" and self.options.shared and "MT" in self.settings.compiler.runtime:
+            raise ConanInvalidConfiguration("Visual Studio build for shared library with MT runtime is not supported")
 
     def source(self):
         tools.get(**self.conan_data["sources"][self.version])
         os.rename("cpr-{}".format(self.version), self._source_subfolder)
 
     def requirements(self):
-        if Version(self.version) < "1.5.0":
-            self.requires("libcurl/7.67.0")
-        else:
-            self.requires("libcurl/7.69.1")
-        if self.options.with_openssl:
-            self.output.warn("OpenSSL support is not stable yet. whoshuu/cpr#31")
+        self.requires("libcurl/{}".format("7.67.0" if not self._supports_openssl else "7.69.1"))
```

Comment:
Uff 😅  This is a clear candidate for the `validate()` function. This is a check to do after all the graph has been generated and resolved, other library in the graph might override `libcurl` version making this conditional totally useless...

`validate()` is not yet available in c3i, but we can write it in the `build()` method:

```python

if not self._supports_openssl and tools.Version(self.deps_cpp_info[“libcurl”].version) > "7.67.0":
    raise ConanInvalidConfiguration

```

Feel free to add it to this PR or not, I know we will revisit lots of recipe with the new `validate()` method.

Reply:
I tested it locally and it does _seem to_ work with other versions which is I why I opt'd to not enforce the version check.

At each release there was a single official supported `libcurl` version which is what you see in the requires.

If consumers what something other than the regression tested they can override.

(that being said, that feature is still a huge need and we do need it for other cases like the `oatpp` family)

Reply:
> At each release there was a single official supported `libcurl` version which is what you see in the requires.

If this is explicitly said and the library works only with a given version of `libcurl` it is something to check (and raise) in the `validate()` method, otherwise in ConanCenter we should try to declare the same version of a requirement everywhere (and probably the only possible agreement is to use the latest one). The user can always override, and it will work, but probably they will get a conflict before (not a good experience) and after overriding it binaries won't be available. Both are value propositions of ConanCenter. 

Anyway, there is some work ahead on our side before considering this an issue 😄